thetis/timeintegrator.py

```diff
@@ -433,6 +433,127 @@ def advance(self, t, update_forcings=None):
+                            fields[k].function_space())
+                    elif isinstance(fields[k], Constant):
+                        self.fields_old[tracer][k] = Constant(fields[k])
+
+        # Handle implicitness
+        if not isinstance(semi_implicit, dict):
+            semi_implicit = {tracer: semi_implicit for tracer in self.solutions}
+        if not isinstance(theta, dict):
+            theta = {tracer: theta for tracer in self.solutions}
+        theta_const = {tracer: Constant(theta[tracer]) for tracer in self.solutions}
+        solutions_nl = {
+            tracer: self.solutions_old[tracer] if semi_implicit[tracer] else self.solutions[tracer]
+            for tracer in self.solutions
+        }
+
+        # Forms
+        self.forms = OrderedDict({
+            tracer:
+                equation.mass_term(self.solutions[tracer]) - equation.mass_term(self.solutions_old[tracer])
+                - self.dt_const*(
+                    theta_const[tracer]*equation.residual(
+                        'all', self.solutions[tracer], solutions_nl[tracer], self.fields[tracer],
+                        self.fields[tracer], bnd_conditions[tracer],
+                    )
+                    + (1-theta_const[tracer])*equation.residual(
+                        'all', self.solutions_old[tracer], self.solutions_old[tracer],
+                        self.fields_old[tracer], self.fields_old[tracer], bnd_conditions[tracer],
+                    )
+                )
+            for tracer, equation in self.equations.items()
+        })
+        self.update_solver()
+
+    def update_solver(self):
+        """Create solver objects"""
+        if self.solver_parameters.get('pc_type') == 'lu':
+            self.solver_parameters['mat_type'] = 'aij'
+        self.solvers = OrderedDict({
+            tracer: NonlinearVariationalSolver(NonlinearVariationalProblem(F, self.solutions[tracer]),
+                                               solver_parameters=self.solver_parameters,
+                                               options_prefix=self.name)
+            for tracer, F in self.forms.items()
+        })
+
+    def initialize(self, solutions):
+        """Assigns initial conditions to all required fields."""
+        for tracer, solution in solutions.items():
+            self.solutions_old[tracer].assign(solution)
+            for k in sorted(self.fields_old[tracer]):
+                self.fields_old[tracer][k].assign(self.fields[tracer][k])
+
+    def advance(self, t, update_forcings=None):
+        """Advances equations for one time step."""
+        if update_forcings is not None:
+            update_forcings(t + self.dt)
+        for tracer, solution in self.solutions.items():
+            self.solutions_old[tracer].assign(solution)
+
+        for it in range(self.iterations):
+            for tracer, solver in self.solvers.items():
+                with timed_stage(f"{tracer} solve"):
+                    solver.solve()
```

Comment:
As we've declared theta>=0.5 in options, this indeed always means a Picard iteration? Do you think it would be possible to somehow monitor convergence? PETSc has these nice monitors which as quite useful in practice.